tree/treeplayer/inc/ROOT/TDataFrame.hxx

```diff
@@ -709,169 +717,131 @@ private:
       }
       // real branch
       auto branchEl = dynamic_cast<TBranchElement *>(branch);
       if (!branchEl) { // This is a fundamental type
-         auto title    = branch->GetTitle();
+         auto title = branch->GetTitle();
          auto typeCode = title[strlen(title) - 1];
-         if (typeCode == 'B') {
-            return SimpleAction<char, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         }
-         // else if (typeCode == 'b') { return SimpleAction<Uchar, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots); }
-         // else if (typeCode == 'S') { return SimpleAction<Short_t, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots); }
-         // else if (typeCode == 's') { return SimpleAction<UShort_t, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots); }
-         else if (typeCode == 'I') {
-            return SimpleAction<int, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         }
-         // else if (typeCode == 'i') { return SimpleAction<unsigned int , ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots); }
-         // else if (typeCode == 'F') { return SimpleAction<float, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots); }
-         else if (typeCode == 'D') {
-            return SimpleAction<double, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         }
-         // else if (typeCode == 'L') { return SimpleAction<Long64_t, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots); }
-         // else if (typeCode == 'l') { return SimpleAction<ULong64_t, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots); }
-         else if (typeCode == 'O') {
-            return SimpleAction<bool, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         }
+         if (typeCode == 'B') { return BuildAndBook<char>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (typeCode == 'I') { return BuildAndBook<int>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (typeCode == 'D') { return BuildAndBook<double>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'b') { return BuildAndBook<UChar_t>(bl, r, nSlots); }
+         // else if (typeCode == 'S') { return BuildAndBook<Short_t>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 's') { return BuildAndBook<UShort_t>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'i') { return BuildAndBook<int>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'F') { return BuildAndBook<float>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'L') { return BuildAndBook<Long64_t>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'l') { return BuildAndBook<ULong64_t>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'O') { return BuildAndBook<bool>(bl, r, nSlots, (AT_t*)nullptr); }
       } else {
          std::string typeName = branchEl->GetTypeName();
```

Comment:
Please consider comparing the TClass returned by TBranch::GetExpectedType.  (and use a static TClassRef to cache the 'target' TClass.